ktor-network/posix/src/io/ktor/network/sockets/DatagramSendChannel.kt

```diff
@@ -0,0 +1,180 @@
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.network.util.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.errors.*
+import io.ktor.utils.io.pool.*
+import kotlinx.atomicfu.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.selects.*
+import kotlinx.coroutines.sync.*
+import platform.posix.*
+
+@SharedImmutable
+private val CLOSED: (Throwable?) -> Unit = {}
+@SharedImmutable
+private val CLOSED_INVOKED: (Throwable?) -> Unit = {}
+
+internal class DatagramSendChannel(
+    val descriptor: Int,
+    val socket: DatagramSocketImpl
+) : SendChannel<Datagram> {
+    private val onCloseHandler = atomic<((Throwable?) -> Unit)?>(null)
+    private val closed = atomic(false)
+    private val closedCause = atomic<Throwable?>(null)
+    private val lock = Mutex()
+
+    @ExperimentalCoroutinesApi
+    override val isClosedForSend: Boolean
+        get() = socket.isClosed
+
+    override fun close(cause: Throwable?): Boolean {
+        if (!closed.compareAndSet(false, true)) {
+            return false
+        }
+
+        closedCause.value = cause
+
+        if (!socket.isClosed) {
+            socket.close()
+        }
+
+        closeAndCheckHandler()
+
+        return true
+    }
+
+    @OptIn(InternalCoroutinesApi::class)
+    override fun trySend(element: Datagram): ChannelResult<Unit> {
+        if (!lock.tryLock()) return ChannelResult.failure()
+
+        var result = false
+
+        try {
+            DefaultDatagramByteBufferPool.useInstance { buffer ->
+                element.packet.copy().readAvailable(buffer)
+
+                val bytes = element.packet.copy().readBytes()
+                var bytesWritten: Int? = null
+                bytes.usePinned { pinned ->
+                    element.address.address.nativeAddress { address, addressSize ->
+                        bytesWritten = sendto(
+                            descriptor,
+                            pinned.addressOf(0),
+                            bytes.size.convert(),
+                            0,
+                            address,
+                            addressSize
+                        ).toInt()
+                    }
+                }
+                when (bytesWritten ?: error("bytesWritten cannot be null")) {
+                    0 -> throw IOException("Failed writing to closed socket")
+                    -1 -> {
+                        if (errno == EAGAIN) {
+                            result = false
+                        } else {
+                            throw PosixException.forErrno()
+                        }
+                    }
+                    else -> result = true
+                }
+            }
+        } finally {
+            lock.unlock()
+        }
+
+        if (result) {
+            element.packet.release()
+        }
+
+        return ChannelResult.success(Unit)
```

Comment:
I am not sure if this should always return success, I just copied this from the JVM implementation. Should this return success even if result is false?

Reply:
I'll check the JVM implementation and will get back with answer